src/gen9_hevc_encoder.c

```diff
@@ -0,0 +1,7376 @@
+
+        if (!surface_priv->scaled_surface_obj[HEVC_SCALED_SURF_4X_ID])
+            return VA_STATUS_ERROR_ALLOCATION_FAILED;
+
+        i965_check_alloc_surface_bo(ctx, surface_priv->scaled_surface_obj[HEVC_SCALED_SURF_4X_ID], 1,
+                                    VA_FOURCC('N', 'V', '1', '2'), SUBSAMPLE_YUV420);
+
+        if (generic_state->b16xme_supported) {
+            i965_CreateSurfaces(ctx,
+                                priv_state->frame_width_16x,
+                                priv_state->frame_height_16x,
+                                VA_RT_FORMAT_YUV420,
+                                1,
+                                &surface_priv->scaled_surface_id[HEVC_SCALED_SURF_16X_ID]);
+            surface_priv->scaled_surface_obj[HEVC_SCALED_SURF_16X_ID] =
+                SURFACE(surface_priv->scaled_surface_id[HEVC_SCALED_SURF_16X_ID]);
+
+            if (!surface_priv->scaled_surface_obj[HEVC_SCALED_SURF_16X_ID])
+                return VA_STATUS_ERROR_ALLOCATION_FAILED;
+
+            i965_check_alloc_surface_bo(ctx, surface_priv->scaled_surface_obj[HEVC_SCALED_SURF_16X_ID], 1,
+                                        VA_FOURCC('N', 'V', '1', '2'), SUBSAMPLE_YUV420);
+        }
+
+        if (generic_state->b32xme_supported) {
+            i965_CreateSurfaces(ctx,
+                                priv_state->frame_width_32x,
+                                priv_state->frame_height_32x,
+                                VA_RT_FORMAT_YUV420,
+                                1,
+                                &surface_priv->scaled_surface_id[HEVC_SCALED_SURF_32X_ID]);
+            surface_priv->scaled_surface_obj[HEVC_SCALED_SURF_32X_ID] =
+                SURFACE(surface_priv->scaled_surface_id[HEVC_SCALED_SURF_32X_ID]);
+
+            if (!surface_priv->scaled_surface_obj[HEVC_SCALED_SURF_32X_ID])
+                return VA_STATUS_ERROR_ALLOCATION_FAILED;
+
+            i965_check_alloc_surface_bo(ctx, surface_priv->scaled_surface_obj[HEVC_SCALED_SURF_32X_ID], 1,
+                                        VA_FOURCC('N', 'V', '1', '2'), SUBSAMPLE_YUV420);
+        }
+
+        if (obj_surface->fourcc == VA_FOURCC_P010) {
+            i965_CreateSurfaces(ctx,
+                                priv_state->frame_width_in_max_lcu,
+                                priv_state->frame_height_in_max_lcu,
+                                VA_RT_FORMAT_YUV420,
+                                1,
+                                &surface_priv->surface_id_nv12);
+            surface_priv->surface_obj_nv12 = SURFACE(surface_priv->surface_id_nv12);
+
+            if (!surface_priv->surface_obj_nv12)
+                return VA_STATUS_ERROR_ALLOCATION_FAILED;
+
+            i965_check_alloc_surface_bo(ctx, surface_priv->surface_obj_nv12, 1,
+                                        VA_FOURCC('N', 'V', '1', '2'), SUBSAMPLE_YUV420);
+        }
+
+        surface_priv->surface_reff = obj_surface;
+
```

Comment:
Please move the below definition earlier to avoid the memory leak when it fails on buffer allocation.